### PR TITLE
time: remove `Box` from `Sleep`

### LINKED ISSUE #3267: time: Sleep should be !Unpin
As a follow up to #3080, the `Sleep` future should be `!Unpin`. This finishes the work of transitioning the timer to the intrusive weaker style.

### PR DESCRIPTION
**Pull request by carllerche**
Removes the box from `Sleep`, taking advantage of intrusive wakers. The
`Sleep` future is now `!Unpin`.

`Interval` boxes the `Sleep` internally. The box is reused across polls which amortized the allocation. It makes the type a bit easier to use. Implementing an allocation-free `Interval` is fairly trivial using `async-stream` and calling `sleep()`.

Most of the PR is updating code to handle `!Unpin` sleep.

Closes #3267

### CODE REVIEW

File: tokio-test/src/io.rs [hunk 1]
```diff
@@ -67,7 +67,7 @@ enum Action {
 struct Inner {
     actions: VecDeque<Action>,
     waiting: Option<Instant>,
-    sleep: Option<Sleep>,
+    sleep: Option<Pin<Box<Sleep>>>,
     read_wait: Option<Waker>,
     rx: mpsc::UnboundedReceiver<Action>,
```
Comment (Darksonn) [inline tokio-test/src/io.rs:72]: Note that this will probably merge conflict with [the stream PR](https://github.com/tokio-rs/tokio/pull/3277).

**Review by Darksonn — APPROVED** [review]
Looks good to me!

File: tokio-test/src/io.rs (hunk 1 above)
Comment (carllerche) [reply tokio-test/src/io.rs:72]: Yeah, I will hold off a bit to see if we can get that one landed first.

File: tokio/tests/time_sleep.rs [hunk 2]
```diff
@@ -317,15 +317,15 @@ async fn drop_after_reschedule_at_new_scheduled_time() {
```
Comment (bdonlan) [inline tokio/tests/time_sleep.rs:317]: We should have a test that asserts that `Sleep` is `!Unpin` (same for other types that contain pinned `Sleep`s).

Comment (Darksonn) [reply tokio/tests/time_sleep.rs:317]: Such a test can easily be constructed using the macros in `tokio/tokio/tests/async_send_sync.rs`.

**Pull request merged**